File: Day1-9/6.py
def serialize(bins):
    return ','.join(map(str, bins))


def cycles_until_reseen(bins):
    configs = set()

    curr = bins
    while serialize(curr) not in configs:
        configs.add(serialize(curr))
        min_idx = 0
        for i in range(1, len(curr)):
            if curr[i] > curr[min_idx]:
                min_idx = i
        redistribute = curr[min_idx]
        curr[min_idx] = 0
        for i in range(redistribute):
            curr[(min_idx+i+1) % len(curr)] += 1
    return len(configs), curr

if __name__ == '__main__':
    with open('6.txt') as f:
        bins = [int(i) for i in f.read().split()]
    num_cycles, last_seen = cycles_until_reseen(bins)
    print("Part 1: {}".format(num_cycles))
    loop_cycles, _ = cycles_until_reseen(last_seen)
    print("Part 2: {}".format(loop_cycles))

File: Day1-9/6.go
package main

import (
	"encoding/binary"
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func serializeBins(bins []int64) string {
	bytes := make([]byte, len(bins)*8)
	for i := 0; i < len(bins); i++ {
		binary.LittleEndian.PutUint64(bytes[i*8:], uint64(bins[i]))
	}
	return string(bytes)
}

func cyclesUntilReseen(bins []int64) (int, []int64) {
	configs := make(map[string]bool)

	for {
		_, contains := configs[serializeBins(bins)]
		if contains {
			break
		}

		configs[serializeBins(bins)] = true
		minIdx := 0
		for i := 1; i < len(bins); i++ {
			if bins[i] > bins[minIdx] {
				minIdx = i
			}
		}

		redistribute := bins[minIdx]
		bins[minIdx] = 0
		allGains := redistribute / int64(len(bins))
		partialGainsBins := int(redistribute % int64(len(bins)))
		for i := 0; i < len(bins); i++ {
			bins[i] += allGains
			if i < partialGainsBins {
				bins[(minIdx+i+1)%len(bins)]++
			}
		}
	}

	return len(configs), bins
}

func main() {
	dat, _ := ioutil.ReadFile("./6.txt")
	strNums := strings.Fields(string(dat))
	bins := make([]int64, len(strNums))
	for i := 0; i < len(strNums); i++ {
		num, _ := strconv.Atoi(strNums[i])
		bins[i] = int64(num)
	}

	cycles, lastSeen := cyclesUntilReseen(bins)
	fmt.Printf("Part 1: %d\n", cycles)
	loopCycles, _ := cyclesUntilReseen(lastSeen)
	fmt.Printf("Part 2: %d\n", loopCycles)
}
